### PR TITLE
Sprint/filedicts/unify functions

### PR DESCRIPTION
**Pull request by denizural**
- [x] adds some inline documentation for advanced users
- [x] unifies path checks
- [x] adds unit tests for new and some other older functions. Increased code coverage. Found bugs in the older code
- [x] multiple failure cases & type checks for function arguments 
- [x] cleans some of the older redundant code.
- [x] `_resolve_paths` --> `_resolve_abs_paths` : better method name
- [ ] TODO: WIP. Convert all bare exceptions to more specific ones

### CODE REVIEW

File: src/esm_runscripts/filedicts.py [hunk 1]
```diff
@@ -314,10 +317,39 @@ def _resolve_paths(self) -> None:
         for key, path in self.locations.items():
             self[f"absolute_path_in_{key}"] = path.joinpath(self[f"name_in_{key}"])
 
-    def _path_type(self, path: pathlib.Path) -> Union[str, bool]:
+    def convert_to_path(self, path: Union[str, pathlib.Path]) -> pathlib.Path:
```
Comment (pgierz) [inline src/esm_runscripts/filedicts.py]: I really don't think we need this function. Have a look:
```
bash-5.1$ python
Python 3.9.5 (default, May 18 2021, 12:31:01)
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pathlib
>>> pathlib.Path(True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/pgierz/opt/miniconda3/lib/python3.9/pathlib.py", line 1072, in __new__
    self = cls._from_parts(args, init=False)
  File "/Users/pgierz/opt/miniconda3/lib/python3.9/pathlib.py", line 697, in _from_parts
    drv, root, parts = self._parse_args(args)
  File "/Users/pgierz/opt/miniconda3/lib/python3.9/pathlib.py", line 681, in _parse_args
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not bool
>>> pathlib.Path(123)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/pgierz/opt/miniconda3/lib/python3.9/pathlib.py", line 1072, in __new__
    self = cls._from_parts(args, init=False)
  File "/Users/pgierz/opt/miniconda3/lib/python3.9/pathlib.py", line 697, in _from_parts
    drv, root, parts = self._parse_args(args)
  File "/Users/pgierz/opt/miniconda3/lib/python3.9/pathlib.py", line 681, in _parse_args
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not int
>>> pathlib.Path("a")
PosixPath('a')
>>> pathlib.Path(pathlib.Path("a"))
PosixPath('a')
>>>
```

File: tests/test_esm_runscripts/test_filedicts.py [hunk 2]
```diff
@@ -159,6 +159,74 @@ def test_filedicts_basics(fs):
...
+    assert path_out == path_in 
+
+    # string goes in, Path goes out
+    path_out = simulation_file.convert_to_path(path_str)
+    assert path_out == path_in
+
+    # garbage in, exception out
+    path_str = 12345
+    with pytest.raises(TypeError):
+        path_out = simulation_file.convert_to_path(path_str)
```
Comment (pgierz) [inline tests/test_esm_runscripts/test_filedicts.py]: Unneeded, see above.

File: tests/test_esm_runscripts/test_filedicts.py [hunk 3]
```diff
@@ -409,47 +477,22 @@ def test_ln_raises_exception_when_target_path_does_not_exist(simulation_file, fs
...
-
-    # Captures output (i.e. the user-friendly error)
-    with Capturing() as output:
-        with pytest.raises(SystemExit) as error:
-            sim_file = esm_runscripts.filedicts.SimulationFile(
-                config, "echam.files.jan_surf"
-            )
 
-    # error needs to occur as the path is not absolute
-    assert any(["ERROR: File Dictionaries" in line for line in output])
```
Comment (pgierz) [inline tests/test_esm_runscripts/test_filedicts.py]: Please do not remove this part. I think @mandresm has more plans for it.

File: src/esm_runscripts/filedicts.py [hunk 4]
```diff
@@ -222,35 +240,20 @@ def ln(self, source: AnyStr, target: AnyStr) -> None:
...
 
+        # general checks
+        self._check_source_and_target(source_path, target_path)
+
         points_to_itself = source_path == target_path
         if points_to_itself:
             err_msg = (
                 f"Unable to create symbolic link: `{source_path}` is linking to itself"
             )
             raise OSError(err_msg)
```
Comment (pgierz) [inline src/esm_runscripts/filedicts.py]: Not needed; `os.symlink` does this already for you.

File: src/esm_runscripts/filedicts.py [hunk 5]
```diff
@@ -130,8 +147,9 @@ def __init__(self, full_config: dict, attrs_address: dict):
         if self["type"] not in ["restart", "outdata"]:
             self["name_in_work"] = self.get("name_in_work", self["name_in_computer"])
 
-        # Complete paths for all possible locations
-        self._resolve_paths()
+        # initialize the locations and complete paths for all possible locations
+        locations = dict.fromkeys(location_keys, None)
```
Comment (pgierz) [inline src/esm_runscripts/filedicts.py]: Is this ever actually used?

File: src/esm_runscripts/filedicts.py [hunk 6]
```diff
@@ -120,6 +120,23 @@ def __init__(self, full_config: dict, attrs_address: dict):
...
+        ]
+
+        # Explanation of the file keys:
+        # -----------------------------
+        # - LOCATION_KEY is one of the strings defined in LOCATION_KEY list
+        #
+        # - name_in<LOCATION_KEY> : file name (without path) in the LOCATION_KEY
+        #   - eg. name_in_work
+        # - absolute_path_in_<LOCATION_KEY> : absolute path in the LOCATION_KEY
+        #   - eg. absolute_path_in_run_tree
```
Comment (pgierz) [inline src/esm_runscripts/filedicts.py]: I would directly put these sorts of things into the object doc-string instead of first as comments and then moving them later. A doc-string can be as long as you need. 

At least how I work, I normally have online documentation open on one side, and my program on the other. Since the online is also cross-linked in sphinx, I can look up other parts of the handbook on the fly, which I cannot do here.

File: src/esm_runscripts/filedicts.py [hunk 7]
```diff
@@ -314,10 +317,39 @@ def _resolve_paths(self) -> None:
...
+        """
+        if isinstance(path, pathlib.Path):
+            return path
+        if isinstance(path, str):
+            return pathlib.Path(path)
+        if not isinstance(path, (str, pathlib.Path)):
+            datatype = type(path).__name__
+            raise TypeError(
+                f"Path ``{path}`` has an incompatible datatype ``{datatype}``. str or pathlib.Path is expected"
+            )
```
Comment (pgierz) [inline src/esm_runscripts/filedicts.py]: If you want to put the actual input the user had given into the exception:
```suggestion
        try:
            return pathlib.Path(path)
        except TypeError:
            raise EsmToolsFileDictError(f"Unable to convert {path} from {type(path)} to a pathlib.Path")
```

File: src/esm_runscripts/filedicts.py [hunk 8]
```diff
@@ -363,26 +403,42 @@ def _check_source_and_target(self, source_path, target_path):
...
+        source_path = self.convert_to_path(source_path)
+        target_path = self.convert_to_path(target_path)
+
+        # Types. Eg. file, dir, link, or None
+        try:
+            source_path_type = self._path_type(source_path)
+            target_path_type = self._path_type(target_path)
+        except TypeError as err:
+            print(err)
+            raise
```
Comment (pgierz) [inline src/esm_runscripts/filedicts.py]: Why? Raise it directly from the function...

**Review by pgierz — CHANGES_REQUESTED** [review]
I think there are still some things that could be cleaned up here :-) but it looks better than the first version and is more in line with what `mv` and `cp` are doing! Hooray!

File: tests/test_esm_runscripts/test_filedicts.py (hunk 3 above)
Comment (denizural) [reply tests/test_esm_runscripts/test_filedicts.py]: I brought it back but still keep the MUT because this is better for unit testing.

File: src/esm_runscripts/filedicts.py (hunk 1 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: I removed the method `convert_to_path` and the test associated with that but now I am getting failing tests because we can still have non `str` or `Path` object and there are no checks covering that.

File: tests/test_esm_runscripts/test_filedicts.py (hunk 3 above)
Comment (pgierz) [reply tests/test_esm_runscripts/test_filedicts.py]: My comment only referred to that line, not the entire method

File: src/esm_runscripts/filedicts.py (hunk 6 above)
Comment (mandresm) [reply src/esm_runscripts/filedicts.py]: agreed

File: src/esm_runscripts/filedicts.py [hunk 9]
```diff
@@ -120,6 +120,23 @@ def __init__(self, full_config: dict, attrs_address: dict):
...
         self.path_in_computer = pathlib.Path(self["path_in_computer"])
 
+        # possible paths for files:
+        location_keys = [
+            "computer",  # pool/source directory (for input files)
+            "exp_tree",  # file in the category directory in experiment directory (eg. input, output, ...)
+            "run_tree",  # file in the experiment/run_<DATE>/<CATEGORY>/ directory
+            "work",  # file in the current work directory. Eg. experiment/run_<DATE>/work/
+        ]
+
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]: I find this redundant. I don't see the need of initializing this variable. And it is not used anywhere since `locations` down in the code are also not used anywhere.

File: src/esm_runscripts/filedicts.py (hunk 5 above)
Comment (mandresm) [reply src/esm_runscripts/filedicts.py]: I don't think this is needed, and it seems redundant to me. All we need is in `self._resolve_paths`

File: tests/test_esm_runscripts/test_filedicts.py (hunk 2 above)
Comment (denizural) [reply tests/test_esm_runscripts/test_filedicts.py]: Removed it

File: src/esm_runscripts/filedicts.py [hunk 10]
```diff
@@ -326,20 +329,25 @@ def _path_type(self, path: pathlib.Path) -> Union[str, bool]:
...
         -------
-        str or bool
+        str or None
             If the path exists it returns its type as a string (``file``, ``dir``,
-            ``link``). If it doesn't exist returns ``False``.
+            ``link``). If it doesn't exist returns ``None``.
         """
-        if path.is_file():
+
+        # TODO: check docstring again
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]: Please, resolve this before merge

File: src/esm_runscripts/filedicts.py (hunk 5 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: Forgot to put the `self.`

File: src/esm_runscripts/filedicts.py (hunk 7 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: Removed it

File: src/esm_runscripts/filedicts.py (hunk 8 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: Just to be more explicit about it. This will reraise the TypeError.

File: src/esm_runscripts/filedicts.py [hunk 11]
```diff
@@ -363,26 +371,38 @@ def _check_source_and_target(self, source_path, target_path):
...
-        if not source_path_type:
-            raise Exception(f"Source file ``{source_path}`` does not exist!")
-        # Target exist
+        # Source does not exist
+        if source_path_type is None:
+            err_msg = f"Unable to perform file operation. Source ``{source_path}`` does not exist!"
+            raise FileNotFoundError(err_msg)
+
+        # Target already exists
+        target_exists = os.path.exists(target_path) or os.path.islink(target_path)
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]: This is not used. Please, remove it

File: src/esm_runscripts/filedicts.py [hunk 12]
```diff
@@ -363,26 +371,38 @@ def _check_source_and_target(self, source_path, target_path):
...
         if target_path_type:
-            # TODO: Change this behavior
-            raise Exception(f"File ``{target_path_type}`` already exists!")
-        # Target dir exists
-        if not target_path_parent_type:
+            err_msg = f"Unable to perform file operation. Target ``{target_path}`` already exists"
+            # TODO: ??? Change this behavior
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]: Remove this TODO when solved

File: src/esm_runscripts/filedicts.py [hunk 13]
```diff
@@ -363,26 +371,38 @@ def _check_source_and_target(self, source_path, target_path):
...
+            raise FileExistsError(err_msg)
+
+        # Target parent directory does not exist
+        if not target_path.parent.exists():
             # TODO: we might consider creating it
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]: I wouldn't do that, it has the potential to create target directories in crazy places if the user makes a mistake on typing the paths. I would only do that in the case of `name_in_{location}` containing a folder. But that would go somewhere else.

File: tests/test_esm_runscripts/test_filedicts.py [hunk 14]
```diff
@@ -159,6 +159,60 @@ def test_filedicts_basics(fs):
...
+    output = simulation_file._path_type(mylink)
+    assert output == "link"
+
+    # check for file
+    output = simulation_file._path_type(simulation_file["absolute_path_in_computer"])
+    assert output == "file"
+
+    # check for directory
+    output = simulation_file._path_type(simulation_file.locations["work"])
+    assert output == "dir"
```
Comment (mandresm) [inline tests/test_esm_runscripts/test_filedicts.py]: Shouldn't we split all of this in different unit tests?

Comment (mandresm) [reply tests/test_esm_runscripts/test_filedicts.py]: test descriptions are missing

File: src/esm_runscripts/filedicts.py (hunk 6 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: Moved them into the docstring with some real examples

File: tests/test_esm_runscripts/test_filedicts.py [hunk 15]
```diff
@@ -159,6 +159,60 @@ def test_filedicts_basics(fs):
...
+    source_path = Path("/this/does/not/exist")
+    target_path = simulation_file.locations["work"]
+    with pytest.raises(FileNotFoundError):
+        simulation_file._check_source_and_target(source_path, target_path)
+
+    # target does not exist
+    source_path = simulation_file.locations["computer"] 
+    target_path = Path("/this/does/not/exist")
+    with pytest.raises(FileNotFoundError):
+        simulation_file._check_source_and_target(source_path, target_path)
```
Comment (mandresm) [inline tests/test_esm_runscripts/test_filedicts.py]: I think this can also be devided in different tests. One can turn the comments into test descriptions once the tests are separated.

File: tests/test_esm_runscripts/test_filedicts.py [hunk 16]
```diff
@@ -409,47 +463,28 @@ def test_ln_raises_exception_when_target_path_does_not_exist(simulation_file, fs
...
-            jan_surf:
-                type: input
-                name_in_computer: T63CORE2_jan_surf.nc
-                name_in_work: unit.24
-                path_in_computer: pool/ECHAM/T63
-        experiment_input_dir: /work/ollie/pgierz/some_exp/input/echam
-        thisrun_input_dir: /work/ollie/pgierz/some_exp/run_20010101-20010101/input/echam
-    """
-    config = yaml.safe_load(dummy_config)
+    simulation_file.path_in_computer = Path("foo/bar")
```
Comment (mandresm) [inline tests/test_esm_runscripts/test_filedicts.py:645]: I'm okay with having a dictionary defined somewhere else in the tests to be used as default, but I don't like that the object is instanced outside of the test. IMO the `SimulationFile` object should be instanced in the test, after we have modified the default `dict` manually on the test too. This method is closer to our real case scenario: 1) user defines dict, 2) ESM-Tools handles that dict. Right now you are doing: 1) test defines dict, 2) test instances dict, 3) test modifies that instance using the object methods, 4) test handles the operation of the test

File: tests/test_esm_runscripts/test_filedicts.py [hunk 17]
```diff
@@ -409,47 +463,28 @@ def test_ln_raises_exception_when_target_path_does_not_exist(simulation_file, fs
...
 
     # Captures output (i.e. the user-friendly error)
     with Capturing() as output:
         with pytest.raises(SystemExit) as error:
-            sim_file = esm_runscripts.filedicts.SimulationFile(
-                config, "echam.files.jan_surf"
-            )
+            simulation_file._check_path_in_computer_is_abs()
```
Comment (mandresm) [inline tests/test_esm_runscripts/test_filedicts.py:650]: same as above

**Review by mandresm — CHANGES_REQUESTED** [review]

File: src/esm_runscripts/filedicts.py (hunk 13 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: Comment is not from me. This is basically a guard against creating that target whose parent does not exist.

Comment (mandresm) [reply src/esm_runscripts/filedicts.py]: True, this comment is mine XD sorry

File: tests/test_esm_runscripts/test_filedicts.py (hunk 14 above)
Comment (denizural) [reply tests/test_esm_runscripts/test_filedicts.py]: Good idea. I wrote the test when I saw that the function was not covered. It is a good idea to refactor the tests and split them when they get large.

File: tests/test_esm_runscripts/test_filedicts.py (hunk 15 above)
Comment (denizural) [reply tests/test_esm_runscripts/test_filedicts.py]: Good idea, I will do it.

File: src/esm_runscripts/filedicts.py (hunk 9 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: This comment is outdated. I already moved it to docstrings.

File: src/esm_runscripts/filedicts.py (hunk 13 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: 🤣 yes, that happens all the time. I think the solution is placing the name initial before the comment 😄

File: src/esm_runscripts/filedicts.py (hunk 11 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: Sorry, that was a typo. I am fixing it.

File: src/esm_runscripts/filedicts.py (hunk 9 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: `self.locations` is now initialized with `None` values

File: src/esm_runscripts/filedicts.py (hunk 10 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: solved

File: tests/test_esm_runscripts/test_filedicts.py (hunk 14 above)
Comment (denizural) [reply tests/test_esm_runscripts/test_filedicts.py]: Newest version has separate test cases for the `_path_type_` method.

File: tests/test_esm_runscripts/test_filedicts.py (hunk 15 above)
Comment (denizural) [reply tests/test_esm_runscripts/test_filedicts.py]: Already implemented and pushed.

File: src/esm_runscripts/filedicts.py [hunk 18]
```diff
@@ -120,6 +147,9 @@ def __init__(self, full_config: dict, attrs_address: dict):
         self.component = component = attrs_address.split(".")[0]
         self.path_in_computer = pathlib.Path(self["path_in_computer"])
 
+        # possible paths for files:
+        location_keys = ["computer", "exp_tree", "run_tree", "work"]
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]: Would it make sense to move this below the `# Initialize the locations and complete paths`? In this way the whole initialization is together and clearer.

File: tests/test_esm_runscripts/test_filedicts.py (hunk 16 above)
Comment (denizural) [reply tests/test_esm_runscripts/test_filedicts.py:645]: I see your point, but the **unit test** aims to test a **specific** function/method behaviour with different edge cases. Testing the behaviour of the object is also important (eg. functional test). I propose to have the following:
- have very specific unit tests for every single method to increase the coverage.
- have buggy YAML files with descriptive names. Eg. `config_no_absolute_paths.yaml` and write **functional tests** that test the higher level behaviour of the object when these YAML files are parsed.

What do you guys think @mandresm @pgierz

File: src/esm_runscripts/filedicts.py [hunk 19]
```diff
@@ -173,12 +204,13 @@ def cp(self, source: str, target: str) -> None:
         source_path = self[f"absolute_path_in_{source}"]
         target_path = self[f"absolute_path_in_{target}"]
 
-        # Checks
+        # general checks
+        # TODO (deniz): need to add higher level exception handler (eg. user_error)
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py:225]:
```suggestion
```
No please, we agreed we will only use user_error for "extended-yaml syntax" problems

File: src/esm_runscripts/filedicts.py [hunk 20]
```diff
@@ -193,7 +225,7 @@ def cp(self, source: str, target: str) -> None:
 
     @_allowed_to_be_missing
     def ln(self, source: AnyStr, target: AnyStr) -> None:
-        """creates symbolic links from the path retrieved by ``source`` to the one by ``target``
+        """creates symbolic links from the path retrieved by ``source`` to the one by ``target``. ``source`` and ``target`` keys will be transformed into their corresponding absolute paths and ``target`` will be created.
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]: I don't understand the extra documentation here. Nothing gets transformed in this function right?

File: src/esm_runscripts/filedicts.py [hunk 21]
```diff
@@ -222,35 +254,14 @@ def ln(self, source: AnyStr, target: AnyStr) -> None:
...
-            err_msg = (
-                f"Unable to create symbolic link: `{source_path}` is linking to itself"
-            )
-            raise OSError(err_msg)
-
-        if not os.path.exists(source_path):
-            err_msg = f"Unable to create symbolic link: source file `{source_path}` does not exist"
-            raise FileNotFoundError(err_msg)
+        # general checks
+        # TODO (deniz): need to add higher level exception handler (eg. user_error)
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py:275]:
```suggestion
```
As above, nothing to be done here

File: src/esm_runscripts/filedicts.py [hunk 22]
```diff
@@ -273,7 +284,8 @@ def mv(self, source: str, target: str) -> None:
         source_path = self[f"absolute_path_in_{source}"]
         target_path = self[f"absolute_path_in_{target}"]
 
-        # Checks
+        # general checks
+        # TODO (deniz): need to add higher level exception handler (eg. user_error)
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py:305]:
```suggestion
```
Same

File: src/esm_runscripts/filedicts.py [hunk 23]
```diff
@@ -362,27 +407,41 @@ def _check_source_and_target(self, source_path, target_path):
...
-        source_path_type = self._path_type(source_path)
-        target_path_type = self._path_type(target_path)
-        target_path_parent_type = self._path_type(target_path.parent)
+        # Types. Eg. file, dir, link, or None
+        try:
+            source_path_type = self._path_type(source_path)
+            target_path_type = self._path_type(target_path)
+        except TypeError as err:
+            print(err)
+            raise
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]: What is the purpose of this change?

File: src/esm_runscripts/filedicts.py [hunk 24]
```diff
@@ -362,27 +407,41 @@ def _check_source_and_target(self, source_path, target_path):
...
+
+        # Target parent directory does not exist
+        if not target_path.parent.exists():
+            # TODO: we might consider creating it (Miguel)
+            err_msg = f"Unable to perform file operation. Parent directory of the target ``{target_path}`` does not exist"
+            raise FileNotFoundError(err_msg)
+
+        # if source is a broken link. Ie. pointing to a non-existing file
+        if source_path_type == FileTypes.BROKEN_LINK:
+            err_msg = f"Unable to create symbolic link: `{source_path}` points to a broken path: {source_path.resolve()}"
```
Comment (mandresm) [inline src/esm_runscripts/filedicts.py]:
```suggestion
            err_msg = f"Unable to create symbolic link: ``{source_path}`` points to a broken path: {source_path.resolve()}"
```

**Review by mandresm — APPROVED** [review]
After the minor changes I suggested I think this is ready to be merged

File: src/esm_runscripts/filedicts.py (hunk 19 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py:225]: I think I had to write a better comment there. I meant that we could wrap this in a `try-except` block and exception would call the `user_error`

File: src/esm_runscripts/filedicts.py (hunk 23 above)
Comment (denizural) [reply src/esm_runscripts/filedicts.py]: It was re-raising the exception. Just removed it now. Tests are still passing
